flink-end-to-end-perf-tests/flink-basic-operations/src/main/java/org/apache/flink/basic/utils/FromElements.java

```diff
@@ -0,0 +1,229 @@
+		public void initializeState(FunctionInitializationContext context) throws Exception {
+			Preconditions.checkState(this.operateState == null,
+				"The " + getClass().getSimpleName() + " has already been initialized.");
+			this.operateState = context.getOperatorStateStore().getOperatorState(
+				new ListStateDescriptor<>("from-elements-state", LongSerializer.INSTANCE));
+			if (context.isRestored()) {
+				List<Long> retrievedStates = new ArrayList<>();
+				for (Long entry : this.operateState.get()) {
+					retrievedStates.add(entry);
+				}
+				this.numElementsToSkip = retrievedStates.get(0);
+			}
+		}
+
+		@Override
+		public void run(SourceContext<T> ctx) throws Exception {
+			ByteArrayInputStream bais = new ByteArrayInputStream(elementsSerialized);
+			final DataInputView input = new DataInputViewStreamWrapper(bais);
+			// if we are restored from a checkpoint and need to skip elements, skip them
+			long toSkip = numElementsToSkip;
+			if (toSkip > 0) {
+				try {
+					while (toSkip > 0) {
+						serializer.deserialize(input);
+						toSkip--;
+					}
+				} catch (Exception e) {
+					throw new IOException("Failed to deserialize an element from the source. If you are using user-defined serialization (Value and Writable types), check the serialization functions.\nSerializer is " + serializer);
+				}
+				this.numElementsEmitted = this.numElementsToSkip;
+			}
+			while (isRunning && recordCount < maxCount) {
+				recordCount++;
+				T next;
+				try {
+					Thread.sleep(sleepNum);
+					if (bais.available() == 0) {
+						bais.reset();
+					}
+					next = serializer.deserialize(input);
+				} catch (Exception e) {
+					throw new IOException("Failed to deserialize an element from the source. If you are using user-defined serialization (Value and Writable types), check the serialization functions.\nSerializer is " + serializer);
+				}
+				synchronized (ctx.getCheckpointLock()) {
+					ctx.collect(next);
+					numElementsEmitted++;
+					sourceTpsMetrics.markEvent();
+				}
+			}
+		}
+
+		@Override
+		public void cancel() {
+			isRunning = false;
+		}
+
+		@Override
+		public void snapshotState(FunctionSnapshotContext context) throws Exception {
+			Preconditions.checkState(this.operateState != null,
+				"The " + getClass().getSimpleName() + " has not been properly initialized.");
+			this.operateState.clear();
+			if (this.numElementsEmitted != 0L) {
+				this.operateState.add(1L);
+			} else {
+				this.operateState.add(0L);
+			}
+		}
+
+		/**
+		 * Verifies that all elements in the collection are non-null, and are of the given class, or a subclass
+		 * thereof.
+		 *
+		 * @param elements The collection to check.
+		 * @param viewedAs The class to which the elements must be assignable to.
+		 * @param <OUT>    The generic type of the collection to be checked.
+		 */
+		public static <OUT> void checkCollection(Collection<OUT> elements, Class<OUT> viewedAs) {
+			for (OUT elem : elements) {
+				if (elem == null) {
```

Comment:
Should be done in the constructor to actually guarantee that this check is performed.